### PR TITLE
feat: add proc macro to reduce overseer mock boilerplate

### PR DESCRIPTION
**Pull request by drahnr**
DRY: Remove some boiler plate code around `AllSubsystems` helper generic in exchange for a small proc macro.

### CODE REVIEW

File: node/overseer/subsystems-gen/tests/ui/err-01-enum.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#![allow(dead_code)]
+
+use polkadot_procmacro_overseer_subsystems_gen::AllSubsystemsGen;
+
+#[derive(Clone, AllSubsystemsGen)]
+enum AllSubsystems<A,B> {
+	A(A),
+
+    #[sticky]
+    B(B),
```
Comment (ordian) [inline node/overseer/subsystems-gen/tests/ui/err-01-enum.rs:10]: sPaCEs!!!11

File: node/overseer/subsystems-gen/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+				}
+			}).collect::<HashSet<Ident>>();
+
+			let strukt_ty = ds.ident;
+
+			if generic_types.is_empty() {
+				return Err(Error::new(strukt_ty.span(), "struct must have at least one generic parameter."))
+			}
+
+			// collect all fields that exist, and all fields that are replacable
```
Comment (ordian) [inline node/overseer/subsystems-gen/src/lib.rs:69]:
```suggestion
			// collect all fields that exist, and all fields that are replaceable
```

File: node/overseer/subsystems-gen/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashSet;
+
+use proc_macro2::TokenStream;
+use quote::quote;
+
+use syn::{AttrStyle, Error, GenericParam, Ident, Result, Type, parse2};
+
+#[proc_macro_derive(AllSubsystemsGen, attributes(sticky))]
```
Comment (ordian) [inline node/overseer/subsystems-gen/src/lib.rs:24]: Are we planning to use `sticky` attribute in `AllSubsystems`? I don't think so (YAGNI).

File: node/overseer/subsystems-gen/tests/ui/err-01-generic-used-twice.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#![allow(dead_code)]
+
+use polkadot_procmacro_overseer_subsystems_gen::AllSubsystemsGen;
+
+#[derive(Clone, AllSubsystemsGen)]
+struct AllSubsystems<X> {
+	a: X,
+    b: X,
```
Comment (ordian) [inline node/overseer/subsystems-gen/tests/ui/err-01-generic-used-twice.rs:8]: ditto

File: node/overseer/subsystems-gen/tests/ui/err-01-no-generic.rs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#![allow(dead_code)]
+
+use polkadot_procmacro_overseer_subsystems_gen::AllSubsystemsGen;
+
+#[derive(Clone, AllSubsystemsGen)]
+struct AllSubsystems {
+	a: f32,
+	#[sticky]
+    b: u16,
```
Comment (ordian) [inline node/overseer/subsystems-gen/tests/ui/err-01-no-generic.rs:9]: same

File: node/overseer/subsystems-gen/tests/ui/err-01-sticky.rs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#![allow(dead_code)]
+
+use polkadot_procmacro_overseer_subsystems_gen::AllSubsystemsGen;
+
+#[derive(Clone, AllSubsystemsGen)]
+struct AllSubsystems<A=(),B=A> {
+	a: A,
+
+    #[sticky]
+    b: B,
```
Comment (ordian) [inline node/overseer/subsystems-gen/tests/ui/err-01-sticky.rs:10]: and this

File: node/overseer/subsystems-gen/tests/ui/err-01-no-generic.rs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#![allow(dead_code)]
+
+use polkadot_procmacro_overseer_subsystems_gen::AllSubsystemsGen;
+
+#[derive(Clone, AllSubsystemsGen)]
+struct AllSubsystems {
+	a: f32,
+	#[sticky]
```
Comment (ordian) [inline node/overseer/subsystems-gen/tests/ui/err-01-no-generic.rs:8]: how this test even passes?
```suggestion
```

File: node/overseer/subsystems-gen/tests/ui/err-01-enum.rs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#![allow(dead_code)]
+
+use polkadot_procmacro_overseer_subsystems_gen::AllSubsystemsGen;
+
+#[derive(Clone, AllSubsystemsGen)]
+enum AllSubsystems<A,B> {
+	A(A),
+
+	#[sticky]
```
Comment (ordian) [inline node/overseer/subsystems-gen/tests/ui/err-01-enum.rs:9]:
```suggestion
```

File: node/overseer/subsystems-gen/tests/ui/err-01-enum.rs [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+	B(B),
+}
+
+fn main() {
+	let all = AllSubsystems::<u8,u16>::A(0u8);
+	// let all = all.replace_a(77u8);
```
Comment (ordian) [inline node/overseer/subsystems-gen/tests/ui/err-01-enum.rs:15]:
```suggestion
```

File: node/overseer/subsystems-gen/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+				let ty_ident = match field.ty {
+					Type::Path(path) => path.path.get_ident().cloned().ok_or_else(|| {
+						Error::new(proc_macro2::Span::call_site(),  "Expected an identifier, but got a path.")
+					}),
+					_ => return Err(Error::new(proc_macro2::Span::call_site(), "Must be path."))
+				}?;
+
+				if generic_types.contains(&ty_ident) {
+					if let Some(previous) = duplicate_generic_detection.replace(ty_ident) {
+						return Err(Error::new(previous.span(), "Generic type parameters may only be used once have at least one generic parameter."))
```
Comment (ordian) [inline node/overseer/subsystems-gen/src/lib.rs:94]: my brain fails to parse this sentence 

also this check could be simplified now that we don't ignore fields, we can check the number of field == number of generics == number of field types

**Review by ordian — APPROVED** [review]
Looks good modulo small nits.

File: node/overseer/subsystems-gen/src/lib.rs (hunk 10 above)
Comment (drahnr) [reply node/overseer/subsystems-gen/src/lib.rs]: Your brain is correct

**Pull request merged**